Std/Tactic/Relation/Symm.lean

```diff
@@ -0,0 +1,128 @@
+/-
+Copyright (c) 2022 Siddhartha Gadgil. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Siddhartha Gadgil, Mario Carneiro, Scott Morrison
+-/
+import Std.Lean.Meta.Basic
+
+/-!
+# `symm` tactic
+
+This implements the `symm` tactic, which can apply symmetry theorems to either the goal or a
+hypothesis.
+-/
+
+set_option autoImplicit true
+
+open Lean Meta
+
+namespace Std.Tactic
+
+/-- Environment extensions for symm lemmas -/
+initialize symmExt :
+    SimpleScopedEnvExtension (Name × Array (DiscrTree.Key true)) (DiscrTree Name true) ←
+  registerSimpleScopedEnvExtension {
+    addEntry := fun dt (n, ks) => dt.insertCore ks n
+    initial := {}
+  }
+
+initialize registerBuiltinAttribute {
+  name := `symm
+  descr := "symmetric relation"
+  add := fun decl _ kind => MetaM.run' do
+    let declTy := (← getConstInfo decl).type
+    let (xs, _, targetTy) ← withReducible <| forallMetaTelescopeReducing declTy
+    let fail := throwError
+      "@[symm] attribute only applies to lemmas proving x ∼ y → y ∼ x, got {declTy}"
+    let some _ := xs.back? | fail
+    let targetTy ← reduce targetTy
+    let .app (.app rel _) _ := targetTy | fail
+    let key ← withReducible <| DiscrTree.mkPath rel
+    symmExt.add (decl, key) kind
+}
+
+end Std.Tactic
+
+open Std.Tactic
+
+namespace Lean.Expr
```

Comment:
All of these functions should be in the `Std.Tactic` namespace, I don't think it's a good idea to create `Expr.symm` here, this is likely to conflict with uses outside std. Ditto for `MVarId.symm`, we don't have a precedent for doing tactic functions with `MVarId` dot notation in std, only lean core does that and I am concerned about name collisions.

Reply:
I guess so. It makes me a little sad to lose the dot notation for both `Expr.symm` and `MVarId.symm`.

Reply:
Is it used ever (outside this file)?

Reply:
`MVarId.symm` is used in one tactic in Mathlib, and `MVarId.rfl` is used in two (`rw?` and `gcongr`). So it's no great loss. I just hadn't realised you wanted to be as cautious about the namespaces for Std; given Std is always going to be near the top of any import hierarchy I'd thought we'd be safe to insert things into core namespaces.

Reply:
I think this is mostly fine, but I would try to limit this to things that are both public API and also not likely to be desired for other purposes either upstream or downstream.

Reply:
Okay, I have changed the incorrect `Mathlib` namespaces to `Std`, and made the `symmAux` functions private.

Reply:
I also slightly prefer a longer name that indicates it's applying a tactic for these (`applySymm`?), but I'll defer to the author on this.

Is there any remaining changes desired before merging?

Reply:
Indeed, the longer name is better. Done now. I'm happy for these to go in.